codex/stores/repostore.nim

```diff
@@ -93,6 +103,17 @@ method getBlock*(self: RepoStore, cid: Cid): Future[?!Block] {.async.} =
   trace "Got block for cid", cid
   return Block.new(cid, data)
 
+proc getBlockExpirationTimestamp(self: RepoStore, ttl: ?Duration): SecondsSince1970 =
+  let duration = ttl |? self.blockTtl
+  self.clock.now() + duration.seconds
```

Comment:
```suggestion
  Moment.fromNow(duration).epochSeconds()
```

Reply:
I would not like to remove the clock member from the repostore. The ability to inject a test-mock is just too convenient. Not directly referencing static library methods from business-logic modules is perhaps a good idea in general. However, I did try and use Moment.now().epochSeconds() instead of std/times in systemclock.nim. As this is the place where I want to wrap the system time. Unfortunately, it seems to be incorrect by a sizable margin. At the moment of running the test for systemclock:
std/times - now.utc as Unix yields 1677060610
chronos - Moment.now().epochSeconds() yields 180628
If the difference was mere hours, timezones could be to blame, but 180628 is January 3 in 1970. I had a look at the chronos code to see what's happening, but I'm afraid the platform-specific details happening there are a bit out of my league.

Reply:
> I did try and use Moment.now().epochSeconds() instead of std/times in systemclock.nim. As this is the place where I want to wrap the system time. Unfortunately, it seems to be incorrect by a sizable margin. At the moment of running the test for systemclock:

Yes, it is indeed broken and the `epoch` methods should have never been exported as the values are opaque and not supposed to be portable... Pity because I hate mixing clocks in this case, it makes things error prone.

> The ability to inject a test-mock is just too convenient.

Mocking _is_ convenient and often overused. In general, mocking makes sense when fine grained control of the different layers is required, otherwise, it's most likely not testing much. 

For example, in the case of the clock, testing for expirations can be done by simply doing an `await sleepAsync(1.seconds)` or similar, and in the case of the repostore, instead of mocking, its possible to initialize the store with in-memory sqlite DBs like this - `SQLiteDatastore.new(Memory).tryGet()` - take a look at how it's done here https://github.com/status-im/nim-datastore/blob/master/tests/datastore/sql/testsqliteds.nim#L18. 

It isn't a big deal, but "two method" classes and other glue should be avoided because it adds unnecessary clutter to the codebase...

Reply:
> or example, in the case of the clock, testing for expirations can be done by simply doing an `await sleepAsync(1.seconds)` or similar,

I agree that mocking is often overused, and using in-memory databases and http servers is often a great way to avoid mocks. But with clocks I don't agree. If you start adding sleeps to your tests, then they all add up when you do a full test run. This slows down the feedback cycle, which slows down development for the entire team.

Reply:
This sleeps can literally be milliseconds, which adds close to nothing to the duration of the test...

Reply:
I think I disagree with both of you.... :D
When a unit-test fails, I want to know for certain that the error is in the module that the test is for. Using real external dependencies (timers, in-memory databases, etc) adds code/behavior to the test. Not only do you have to keep track of this while writing the test, (writing around the behavior of these modules to get them to act how you want for your test) but changes to modules _other_ than the unit under test can cause a unit test to fail. Worse still: If a faulty change is made to a fundamental module of the system that is never mocked out in testing, many or all tests can fail. This removes the use of tests for locating faults in the codebase.
Now you might say the in-memory-DB, Key.init, Cid.init, etc are external libraries and so they are unlikely to cause trouble in this way. Fair enough. But we are building direct dependencies here, and that will make the code (test-code or not) more fragile.
In unit tests, ideally, I want everything _except_ the unit under test to be fake, completely under control of the test, and in no way responsive to changes elsewhere in the codebase. (Date-types not withstanding.)

I know some people object to this one-to-one unit testing approach by saying tests of this sort will do nothing to safeguard the functionality of the total system. This is true, but that's what integration tests are for.
</rant>

I am perfectly happy to set my opinions aside and rolling with the codebase as it's been going. (As you see, I've used the in-memoryDB myself in this changeset.) But here you have my take on the topic.

Reply:
Thanks for sharing your point of view! I'd say your reasoning is uncontroversial and pretty common, if anything, Mark's and my takes are probably the controversial ones. Never the less, let me provide some counterpoints for contrast.

> When a unit-test fails, I want to know for certain that the error is in the module that the test is for.

Yes, but it is a balancing act between "coercing your code into doing something" vs "actually making sure that the code does what you want". Presumably, what we want is the latter.

> Using real external dependencies (timers, in-memory databases, etc) adds code/behavior to the test. Not only do you have to keep track of this while writing the test, (writing around the behavior of these modules to get them to act how you want for your test) but changes to modules _other_ than the unit under test can cause a unit test to fail.

What your describing here, is closer to how the application is going to "exist" in the wild vs our idealized view of that. 

But closer to your point, yes, mocking is usefull when you don't have well behaved modules or when you need more control than you otherwise get from these modules. In other words, mocking is a sort of necessary evil, even by your own account - "writing around the behavior of these modules to get them to act how you want for your test".

But more broadly, mocking is a usefull tool, but as every tool it has a time and a place.

> Worse still: If a faulty change is made to a fundamental module of the system that is never mocked out in testing, many or all tests can fail.

For all intents and purposes, this is good! You want your test to fail when the underlying functionality has changed in some fundamental way.

> This removes the use of tests for locating faults in the codebase.

I would argue that this extends beyond the surface of the codebase and improves the reach of testing.

> Now you might say the in-memory-DB, Key.init, Cid.init, etc are external libraries and so they are unlikely to cause trouble in this way.

Thats never the point, the point is to be able to locate and detect errors, be it in code directly under the control of the project or any third party modules it relies on. This is a fundamental point, and I'm willing to concede that what we do is **not** unit testing, in the strict cense of what unitesting means, but I would also argue, that the usefulness of unitesting is quite limited if looked at objectively.

> But we are building direct dependencies here, and that will make the code (test-code or not) more fragile.

In a way we are and it is intentionall. Partly because of the nature of the software we're building and partly due to the very well understood limitations of unittests and their innadequacy in many scenarios.

So what do we actually do? What is our (so far informal) testing approach? I would characterize it as mostly bottom-up testing vs top-down testing. With top-down testing being more akin to unittesting.

It mostly relates to where the complexity of the software lies and what is the most convenient way to build it? Is it top down or bottom up? 

If the complexity is mostly at the surface, most likely it would start being build top down. This is where mocking is usually prevalent, a mock is built for the components bellow the ones being tested as you keep descending down the component tree. Unsurprisingly this is usually software with complex user interactions, often complicated UIs or similar - sales, trading, medical, etc...

Codex on the other hand, has very little complexity at the top, most of the core functionality sits behind two calls/methods - `upload` and `download`. As you descend, you get more and more complexity. How you deal with that depends, but its common to start either somewhere in the middle of the component tree or at the very bottom... Once you start building it from the bottom up, most of the components naturally complement each other as you keep ascending and, there is rarely a need for mocking, except for a few narrow cases...

> I know some people object to this one-to-one unit testing approach by saying tests of this sort will do nothing to safeguard the functionality of the total system.

I've seen this, but I've never heard convincing enough arguments for why that is, I think unittesting is perfectly fine, but it cannot be treated as the one and only way of testing correctness and complettness in isolation. I think it all depends on the nature of the software, where complexity lies and how it gets built - as per the arguments above.

> that's what integration tests are for.

Indeed, bottom-up testing is a sort of integration testing, because it allows you to integrate the system as the components are being build from the bottom up, but it is fine grained enough that it allows you to do that with enough control over the flow that it suffices to catch and isolate issues without too much hassle.

Our other integration tests, test the entire system as a whole, with all the components plugged in and an environment as close to a real setup as possible.

I hope this gives a good enough rationale for why things are the way they are :)

Reply:
I understand, thank you. I don't think we're going to agree on this point, but luckily, we don't have to in order to work well together. When I'm writing code, I always ask myself how much knowledge is in the function or module I'm working on. If it has a lot of knowledge about the layout of many parts of the system, the code's going to be very fragile and will need constant tinkering to keep up with changes elsewhere in the codebase. I'm always looking for way to hide away and abstract out knowledge from my modules, to make them make sense in an isolated way, so to speak. I know I have had some respectful disagreements on where to draw this line with.... most of my team members in the past. But as I said, I shall in no way insist on doing things my way.